scraper/src/libretexts2zim/css.py

```diff
@@ -0,0 +1,140 @@
+from collections.abc import Iterable
+from pathlib import Path
+from urllib.parse import urljoin, urlparse
+
+from tinycss2 import ast, parse_stylesheet_bytes, serialize  # pyright: ignore
+from tinycss2.serializer import serialize_url  # pyright: ignore
+
+OriginalUrl = str
+FullZimPath = Path
+RelativeCssPath = Path
+
+
+class CssProcessor:
+    """Utility to to process CSS, extract assets and rewrite URLs
+
+    This utility can process multiple CSS documents that will be stored in a ZIM
+    It extracts the list of assets (images, fonts) that are used in the CSS documents
+    and compute appropriate ZIM paths for each of them.
+
+    Arguments:
+      css_target_path: "folder" where the CSS documents that will be processed will be
+        stored in the ZIM
+      css_assets_root_path: "folder" where the CSS assets referenced in the CSS
+        documents will be stored in the ZIM
+    """
+
+    def __init__(
+        self,
+        css_target_path: Path = Path("/content"),
+        css_assets_root_path: Path = Path("/content/css_assets"),
+    ) -> None:
+        self.css_target_path = css_target_path
+        self.css_assets_root_path = css_assets_root_path
+        self.css_assets: dict[OriginalUrl, FullZimPath] = {}
+        self.used_paths: list[RelativeCssPath] = []
+
+    def process(self, css_original_url: str, css_content: bytes) -> str:
+        rules, _ = parse_stylesheet_bytes(  # pyright: ignore[reportUnknownVariableType]
+            css_content
+        )
+        self._process_list(
+            css_original_url,
+            rules,  # pyright: ignore[reportUnknownArgumentType]
+        )
+        return serialize(rules)
+
+    def _process_url(
+        self, css_original_url: str, css_url: str
+    ) -> RelativeCssPath | None:
+        original_url = urljoin(css_original_url, css_url)
+        original_url_parsed = urlparse(original_url)
+        if original_url_parsed.scheme.lower() not in ["http", "https"]:
+            return None
+        if original_url in self.css_assets:
+            return self.css_assets[original_url].relative_to(self.css_target_path)
+        original_path = Path(urlparse(original_url).path)
+        target_parent = Path(
+            *[
+                parent.name
+                for parent in reversed(original_path.parents)
+                if parent.name and parent.name != ".."
```

Comment:
What's this doing exactly? What's the consequence?

Reply:
This allows to compute a "clean" target path to use inside the ZIM, even if original path was a relative one. E.g. for an asset located at `../folder/file.ext`, it will place the file in the ZIM at `folder/file.ext`. This is meant to create a folder structure "similar" to the online one, but simplifying the file path to get rid of query strings for instance. I will isolate this logic in a dedicated function to add a docstring and make it clearer.

Reply:
My concern is that we are removing those walking paths instead of resolving them so something like `folder/sub1/sub2/sub3/sub4/../../../file.txt` will end up at `folder/sub1/sub2/sub3/sub4/file.ext` and not at `folder/file.ext`.

Given all resources in-file are rewritten with full path, it should not have an impact but, as you suggested, this should all be properly documented as this would lead a future dev into a false trail on some random css issue.